File: examples/modules/filter-example/src/filter/counter.lua
-- example.filter.counter
-- Count objects passed through
--   local filter = require("example.filter.counter").new()
--   input:receiver(filter)
--   filter:receiver(output)
--   print(filter:count())
--
-- An example filter module that will count objects it sees.
module(...,package.seeall)

require("dnsjit.core.log")
require("dnsjit.core.receiver_h")
require("dnsjit.core.producer_h")

local loader = require("dnsjit.core.loader")
loader.load("example-filter-counter/counter")

local ffi = require("ffi")
ffi.cdef[[
typedef struct filter_counter {
    core_log_t      _log;
    core_receiver_t recv;
    void*           ctx;
    core_producer_t prod;
    void*           prod_ctx;

    size_t count;
} filter_counter_t;

core_log_t* filter_counter_log();

void filter_counter_init(filter_counter_t* self);
void filter_counter_destroy(filter_counter_t* self);

core_receiver_t filter_counter_receiver();
core_producer_t filter_counter_producer(filter_counter_t* self);
]]
local C = ffi.C

local t_name = "filter_counter_t"
local filter_counter_t = ffi.typeof(t_name)
local Counter = {}

-- Create a new Counter filter.
function Counter.new()
    local self = {
        _receiver = nil,
        _producer = nil,
        obj = filter_counter_t(),
    }
    C.filter_counter_init(self.obj)
    ffi.gc(self.obj, C.filter_counter_destroy)
    return setmetatable(self, { __index = Counter })
end

-- Return the Log object to control logging of this instance or module.
function Counter:log()
    if self == nil then
        return C.filter_counter_log()
    end
    return self.obj._log
end

-- Return the C functions and context for receiving objects.
function Counter:receive()
    return C.filter_counter_receiver(), self.obj
end

-- Set the receiver to pass objects to.
function Counter:receiver(o)
    self.obj.recv, self.obj.ctx = o:receive()
    self._receiver = o
end

-- Return the C functions and context for producing objects.
function Counter:produce()
    return C.filter_counter_producer(self.obj), self.obj
end

-- Set the producer to get objects from.
function Counter:producer(o)
    self.obj.prod, self.obj.prod_ctx = o:produce()
    self._producer = o
end

-- Return counted objects.
function Counter:count()
    return tonumber(self.obj.count)
end

return Counter

File: examples/modules/filter-example/src/filter/counter.c
#include "config.h"

#include <dnsjit/core/assert.h>
#include <dnsjit/core/log.h>
#include <dnsjit/core/receiver.h>
#include <dnsjit/core/producer.h>

typedef struct filter_counter {
    core_log_t _log;

    core_receiver_t recv;
    void*           ctx;

    core_producer_t prod;
    void*           prod_ctx;

    size_t count;
} filter_counter_t;

static core_log_t       _log      = LOG_T_INIT("example.filter.counter");
static filter_counter_t _defaults = {
    LOG_T_INIT_OBJ("example.filter.counter"),
    0, 0,
    0, 0,
    0
};

core_log_t* filter_counter_log()
{
    return &_log;
}

void filter_counter_init(filter_counter_t* self)
{
    mlassert_self();

    *self = _defaults;
}

void filter_counter_destroy(filter_counter_t* self)
{
    mlassert_self();
}

static void _receive(filter_counter_t* self, const core_object_t* obj)
{
    mlassert_self();
    lassert(obj, "obj is nil");

    if (!self->recv) {
        lfatal("no receiver set");
    }
    self->count++;
    self->recv(self->ctx, obj);
}

core_receiver_t filter_counter_receiver()
{
    return (core_receiver_t)_receive;
}

static const core_object_t* _produce(filter_counter_t* self)
{
    const core_object_t* obj;
    mlassert_self();

    obj = self->prod(self->prod_ctx);
    if (obj) {
        self->count++;
    }

    return obj;
}

core_producer_t filter_counter_producer(filter_counter_t* self)
{
    mlassert_self();

    if (!self->prod) {
        lfatal("no producer set");
    }

    return (core_producer_t)_produce;
}
